README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

## currency_decoration

A package that provides an easy way to format currency with a custom style.



## Preview
![currency_decoration_2](https://user-images.githubusercontent.com/16219419/211183695-7e3b2a44-e24a-4e05-a161-3629e1cde033.png)

## Installation
Add `currency_decoration` as a [dependency in your pubspec.yaml file](https://docs.flutter.dev/development/packages-and-plugins/using-packages).

```yaml  
dependencies:
    currency_decoration: ^1.0.1+1
```  



## Usage

```dart
// Example:
CurrencyDecoration(
    symbol: CurrencySymbol.USD,
    amount: 28945.6778,
    symbolSeparator: ' ',
    symbolAlign: SymbolAlign.left,
    fractionDigits: 2,
    currencyValuePlaceStyle: CurrencyValuePlaceStyle.style2,
    primaryTextStyle: const TextStyle(
        color: Colors.green,
        fontWeight: FontWeight.bold,
        fontSize: 18.0),
    secondaryTextStyle: TextStyle(
        color: Colors.green.shade800,
        fontWeight: FontWeight.bold,
        fontSize: 14.0),
    symbolTextStyle: const TextStyle(
      fontSize: 20.0,
      fontWeight: FontWeight.bold,
    ),
  )
```

## compact
```dart
CurrencyDecoration(
    symbol: CurrencySymbol.USD,
    amount: 134000,
    compact: true,
    currencyValuePlaceStyle: CurrencyValuePlaceStyle.style2,
    primaryTextStyle: TextStyle(
        color: Colors.deepOrangeAccent,
        fontWeight: FontWeight.bold,
        fontSize: 18.0),
  )
```
Output: `$134K`

## Supported default currency symbol

[Link](https://github.com/panand8145/currency_decoration/blob/main/lib/currency_symbol.dart)

## Money format style
### International style
```dart
CurrencyDecoration(
    symbol: CurrencySymbol.USD,
    amount: 134000,
    currencyValuePlaceStyle: CurrencyValuePlaceStyle.style1,
  )
```
Output: `$134,000.00`

### Indian style
```dart
CurrencyDecoration(
    symbol: CurrencySymbol.INR,
    amount: 134000,
    symbolSeparator: '',
    currencyValuePlaceStyle: CurrencyValuePlaceStyle.style2,
  )
```
Output: `₹1,34,000.00`

## Currency Symbol Align

```dart
CurrencyDecoration(
    symbol: CurrencySymbol.USD,
    amount: 134000,
    fractionDigits: 0,
    symbolAlign: SymbolAlign.left,  // output: `$ 134,000`
 // symbolAlign: SymbolAlign.right, // output: `134,000 $`
    symbolSeparator: ' ',
    currencyValuePlaceStyle: CurrencyValuePlaceStyle.style1,
  )
```


## Parameters
| Parameter | Description |
|---|---|
| `key` | Controls widget in the tree. |
| `amount`* | decimal/integer value |
| `symbol`* | currency symbol [Link](https://github.com/panand8145/currency_decoration/blob/main/lib/currency_symbol.dart), also accept any String value like `USD` or `INR` etc. |
| `symbolAlign` | Alignment of symbol default`SymbolAlign.left` |
| `thousandSeparator` | Thousand separator. e.g. 1,000,000 (`','`) or 1.000.000 (`'.'`). It can be set to any desired [String]. |
| `decimalSeparator` | Decimal separator. e.g. 9.10 (`'.'`) or 9,10 (`','`). It can be set to any desired [String]. |
| `symbolSeparator` | Character(s) between the amount and the currency symbol. e.g. $ 9.10 (`' '`) or $9.10 (`''`) |
| `enforceDecimals` | If [enforceDecimals] is set to `true`, decimals will be shown even if it is an integer. e.g. `'$ 5.00'` instead of `'$ 5'`. |
| `showThousandSeparator` | Thousand separator. e.g. 1,000,000 (`','`) or 1.000.000 (`'.'`). It can be set to any desired [String]. |
| `compact` | If [compact] is `true` the compact form will be used. e.g. `'$ 1.23K'` instead of `'$ 1,230'`. |
| `primaryTextStyle` | Text style apply on whole units/value (e.g. 52470.596 then text style apply on 52470). if a primary style does not define then use  `DefaultTextStyle.of(context)` as a default style |
| `secondaryTextStyle` | Text style apply on fractional part (e.g. 52470.596 then text style apply on 596). if a secondary style does not define then use `primaryTextStyle` as default |
| `symbolTextStyle` | Text style apply on currency Symbol. if symbol style does not define then use `primaryTextStyle` as default |
| `currencyValuePlaceStyle` | Whole units/value comma style |
| `fractionalStyle` | Style on fractional part/value |
| `fractionDigits` | fractional value |

Parameters marked with * are mandatory to give value.

## MIT License

```
Copyright (c) 2023 Anand Patel

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md

## 1.0.1+2
* README.md file updated

## 1.0.1+1
* README.md file updated

## 1.0.1
* Licence updated

## 1.0.0
* set International currency formatter as a default.

## 0.0.1

* Initial development release

test/currency_decoration_pk_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:currency_decoration_pk/currency_decoration_pk.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/lib/main.dart
import 'package:currency_decoration/currency_decoration.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Currency decoration',
      debugShowCheckedModeBanner: false,
      theme: ThemeData.dark(),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: SingleChildScrollView(
          child: Container(
            margin: const EdgeInsets.all(20.0),
            child: Column(
              children: [
                example_1(context),
                const SizedBox(
                  height: 70.0,
                ),
                const CurrencyDecoration(
                  symbol: CurrencySymbol.USD,
                  amount: 134000,
                  compact: true,
                  symbolAlign: SymbolAlign.left,
                  fractionDigits: 2,
                  enforceDecimals: false,
                  currencyValuePlaceStyle: CurrencyValuePlaceStyle.style2,
                  primaryTextStyle: TextStyle(
                      color: Colors.deepOrangeAccent,
                      fontWeight: FontWeight.bold,
                      fontSize: 18.0),
                ),
                const SizedBox(
                  height: 50.0,
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: const [
                    CurrencyDecoration(
                      symbol: CurrencySymbol.USD,
                      amount: 746722,
                      enforceDecimals: false,
                      currencyValuePlaceStyle: CurrencyValuePlaceStyle.style2,
                      fractionalStyle: FractionalStyle.cross,
                      primaryTextStyle: TextStyle(
                          color: Colors.pinkAccent,
                          fontWeight: FontWeight.bold,
                          fontSize: 18.0),
                    ),
                    SizedBox(
                      width: 20.0,
                    ),
                    CurrencyDecoration(
                      symbol: "Rs.",
                      amount: 79.99,
                      symbolAlign: SymbolAlign.left,
                      fractionDigits: 2,
                      enforceDecimals: false,
                      currencyValuePlaceStyle: CurrencyValuePlaceStyle.style2,
                      primaryTextStyle: TextStyle(
                          color: Colors.lime,
                          fontWeight: FontWeight.bold,
                          fontSize: 18.0),
                    ),
                  ],
                ),
                const SizedBox(
                  height: 50.0,
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: const [
                    CurrencyDecoration(
                      symbol: CurrencySymbol.USD,
                      amount: 99.09,
                      currencyValuePlaceStyle: CurrencyValuePlaceStyle.style1,
                      fractionalStyle: FractionalStyle.cross,
                      primaryTextStyle: TextStyle(
                          color: Colors.deepOrangeAccent,
                          fontWeight: FontWeight.bold,
                          fontSize: 18.0),
                      secondaryTextStyle: TextStyle(fontSize: 14.0),
                    ),
                    SizedBox(
                      width: 20.0,
                    ),
                    CurrencyDecoration(
                      symbol: CurrencySymbol.CNY,
                      amount: 746722,
                      symbolAlign: SymbolAlign.left,
                      fractionDigits: 0,
                      enforceDecimals: false,
                      currencyValuePlaceStyle: CurrencyValuePlaceStyle.style2,
                      primaryTextStyle: TextStyle(
                          color: Colors.white,
                          fontWeight: FontWeight.bold,
                          fontSize: 18.0),
                    ),
                  ],
                ),
                const SizedBox(
                  height: 70.0,
                ),
                Row(
                  children: [
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          ClipRRect(
                            borderRadius: BorderRadius.circular(10.0),
                            child: SizedBox.fromSize(
                              child: Image.network(
                                  "https://cdn.shopify.com/s/files/1/0666/2081/7620/products/CFSC4024BEIGE-1-_1_400x.jpg?v=1665729610",
                                  fit: BoxFit.cover),
                            ),
                          ),
                          const SizedBox(
                            height: 10.0,
                          ),
                          const Text(
                            "FORT COLLINS",
                            style: TextStyle(
                                fontSize: 14.0,
                                color: Colors.white,
                                fontWeight: FontWeight.w500),
                          ),
                          const SizedBox(
                            height: 5.0,
                          ),
                          const Text(
                            "Striped Slim Fit Shirt\n",
                            style: TextStyle(
                                fontSize: 12.0, color: Colors.white60),
                          ),
                          const SizedBox(
                            height: 10.0,
                          ),
                          CurrencyDecoration(
                            symbol: CurrencySymbol.INR,
                            amount: 699.99,
                            symbolAlign: SymbolAlign.left,
                            fractionDigits: 2,
                            enforceDecimals: false,
                            currencyValuePlaceStyle:
                                CurrencyValuePlaceStyle.style2,
                            primaryTextStyle: const TextStyle(
                                color: Colors.blueAccent,
                                fontWeight: FontWeight.w600,
                                fontSize: 16.0),
                            secondaryTextStyle: TextStyle(
                                color: Colors.blueAccent.withOpacity(0.8),
                                fontWeight: FontWeight.w600,
                                fontSize: 12.0),
                          ),
                        ],
                      ),
                    ),
                    const SizedBox(
                      width: 15.0,
                    ),
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          ClipRRect(
                            borderRadius: BorderRadius.circular(10.0),
                            child: SizedBox.fromSize(
                              child: Image.network(
                                  "https://cdn.shopify.com/s/files/1/0666/2081/7620/products/CFDH3066BROWN_1_700x.jpg?v=1665480558",
                                  fit: BoxFit.cover),
                            ),
                          ),
                          const SizedBox(
                            height: 10.0,
                          ),
                          const Text(
                            "DENNISLINGO PREMIUM",
                            style: TextStyle(
                                fontSize: 14.0,
                                color: Colors.white,
                                fontWeight: FontWeight.w500),
                          ),
                          const SizedBox(
                            height: 5.0,
                          ),
                          const Text(
                            "Full Sleeves Slim Fit Classic Shirt",
                            style: TextStyle(
                                fontSize: 12.0, color: Colors.white60),
                          ),
                          const SizedBox(
                            height: 10.0,
                          ),
                          CurrencyDecoration(
                            symbol: CurrencySymbol.USD,
                            amount: 29.99,
                            symbolAlign: SymbolAlign.left,
                            fractionDigits: 2,
                            enforceDecimals: false,
                            currencyValuePlaceStyle:
                                CurrencyValuePlaceStyle.style2,
                            primaryTextStyle: const TextStyle(
                                color: Colors.blueAccent,
                                fontWeight: FontWeight.w600,
                                fontSize: 16.0),
                            secondaryTextStyle: TextStyle(
                                color: Colors.blueAccent.withOpacity(0.8),
                                fontWeight: FontWeight.w600,
                                fontSize: 12.0),
                          ),
                        ],
                      ),
                    ),
                  ],
                )
              ],
            ),
          ),
        ),
      ),
    );
  }

  Widget example_1(BuildContext context) {
    return Container(
      decoration: const BoxDecoration(
        color: Colors.black26,
        borderRadius: BorderRadius.all(
          Radius.circular(8.0),
        ),
      ),
      padding: const EdgeInsets.all(15.0),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const Text(
                "Payable amount is",
                style: TextStyle(color: Colors.white54, fontSize: 11.0),
              ),
              const SizedBox(
                height: 6.0,
              ),
              CurrencyDecoration(
                symbol: CurrencySymbol.USD,
                amount: 28945.6778,
                symbolSeparator: ' ',
                symbolAlign: SymbolAlign.left,
                fractionDigits: 2,
                currencyValuePlaceStyle: CurrencyValuePlaceStyle.style2,
                primaryTextStyle: const TextStyle(
                    color: Colors.green,
                    fontWeight: FontWeight.bold,
                    fontSize: 18.0),
                secondaryTextStyle: TextStyle(
                    color: Colors.green.shade800,
                    fontWeight: FontWeight.bold,
                    fontSize: 14.0),
                symbolTextStyle: const TextStyle(
                  fontSize: 20.0,
                  fontWeight: FontWeight.bold,
                ),
              ),
            ],
          ),
          TextButton(
              onPressed: () {},
              style: ButtonStyle(
                  backgroundColor: MaterialStateColor.resolveWith(
                      (states) => Colors.black45)),
              child: Text(
                "  Pay Now  ",
                style: TextStyle(
                  color: Colors.pink.shade600,
                ),
              )),
        ],
      ),
    );
  }
}


lib/currency_symbol.dart
/// Enumeration for the three possibilities when writing the currency symbol.
enum SymbolAlign { left, right }

enum FractionalStyle { none, strike, cross, disable, ignore }

/// e.g. style1:   i/p 1552470.596   o/p 1,552,470.596
/// e.g. style2:   i/p 1552470.596   o/p 15,52,470.596
enum CurrencyValuePlaceStyle { none, style1, style2 }

class CurrencySymbol {
// none: remove symbol
  static const String none = "";

// Afghanistani Afghani
  static const String AFN = "؋";

// Armenian Dram
  static const String AMD = "դր";

// Azerbaijani Manat
  static const String AZN = "₼";

// Bahraini Dinar
  static const String BHD = ".د.ب";

// Bangladeshi Taka
  static const String BDT = "৳";

// Cambodian Riel
  static const String KHR = "៛";

// Chinese Yuan Renminbi
  static const String CNY = "元";

// Georgian Lari
  static const String GEL = "ლ";

// Indian Rupee
  static const String INR = "₹";

// Iranian Rial
  static const String IRR = "﷼";

// Iraqi Dinar
  static const String IQD = "ع.د";

// Israeli New Sheqel
  static const String ILS = "₪";

// Japanese Yen
  static const String JPY = "¥";

// Jordanian Dinar
  static const String JOD = "د.ا";

// Kazakhstani Tenge
  static const String KZT = "лв";

// Kuwaiti Dinar
  static const String KWD = "د.ك";

// Kyrgyzstani Som
  static const String KGS = "лв";

// Lao Kip
  static const String LAK = "₭";

// Mongolian Tugrik
  static const String MNT = "₮";

// North Korean Won
  static const String KPW = "₩";

// Omani Rial
  static const String OMR = "﷼";

// Philippine Peso
  static const String PHP = "₱";

// Qatari Riyal
  static const String QAR = "﷼";

// Russian Ruble
  static const String RUB = "₽";

// Saudi Arabian Riyal
  static const String SAR = "﷼";

// Korean Won
  static const String KRW = "₩";

// Thai Baht
  static const String THB = "฿";

// United States Dollar
  static const String USD = "\$";

// Turkish New Lira
  static const String TRY = "₺";

// United Arab Emirates Dirham
  static const String AED = "د.إ";

// Uzbekistani Som
  static const String UZS = "лв";

// Viet Nam Dong
  static const String VND = "₫";

// Yemeni Rial
  static const String YER = "﷼";

// Algerian Dinar
  static const String DZD = "دج";

// Eritrean Nakfa
  static const String ERN = "ናቕፋ";

// Ethiopian Birr
  static const String ETB = "ብር";

// Libyan Dinar
  static const String LYD = "ل.د";

// Nigerian Naira
  static const String NGN = "₦";

// Tunisian Dinar
  static const String TND = "د.ت";

// European Euro
  static const String EUR = "€";

// Bulgarian Lev
  static const String BGN = "лв";

// Macedonian Denar
  static const String MKD = "ден";

// Maltese Lira
  static const String MTL = "₤";

// Polish Zloty
  static const String PLN = "zł";

// Serbian Dinar
  static const String RSD = "Дин.";

// Ukrainian Hryvnia
  static const String UAH = "₴";

// United Kingdom Pound Sterling
  static const String GBP = "£";

// Costa Rican Colon
  static const String CRC = "₡";
}


lib/currency_formatter.dart
library currency_formatter;

import 'package:intl/intl.dart' as intl;

import 'currency_symbol.dart';

abstract class CurrencyFormatter {
  static final Map<num, String> _letters = {
    1000000000000: 'T',
    1000000000: 'B',
    1000000: 'M',
    1000: 'K'
  };

  /// Format [amount] using the [settings] of a currency.
  ///
  /// If [compact] is `true` the compact form will be used. e.g. `'$ 1.23K'` instead of `'$ 1,230'`.
  ///
  /// [decimal] is the number of decimal places used when rounding.
  /// e.g. `'$ 45.5'` (1 decimal place), `'$ 45.52'` (2 decimal places).
  ///
  /// If [showThousandSeparator] is set to `false`, the thousand separator won't be shown.
  /// e.g. `'$ 1200'`instead of `'$ 1,200'`.
  ///
  /// If [enforceDecimals] is set to `true`, decimals will be shown even if it is an integer.
  /// e.g. `'$ 5.00'` instead of `'$ 5'`.
  ///
  /// Currency value place [currencyPlaceStyle] is style how currency look like
  /// e.g. style1:   i/p 1552470.596   o/p 1,552,470.596
  /// e.g. style2:   i/p 1552470.596   o/p 15,52,470.596
  ///
  static List<String> format(amount, CurrencyFormatterSettings settings,
      {bool compact = false,
      int fractionDigits = 2,
      showThousandSeparator = true,
      currencyPlaceValueStyle,
      enforceDecimals = false}) {
    amount = double.parse('$amount');
    String numberUnit = "$amount";
    String letter = '';
    String decimalUnit = '';

    if (compact) {
      for (int i = 0; i < _letters.length; i++) {
        if (amount.abs() >= _letters.keys.elementAt(i)) {
          letter = _letters.values.elementAt(i);
          amount /= _letters.keys.elementAt(i);
          break;
        }
      }
      numberUnit = amount.toStringAsPrecision(3);
      numberUnit = numberUnit.replaceAll('.', settings.decimalSeparator!);
    } else {
      if (!showThousandSeparator) {
        settings.thousandSeparator = "";
      }

      if (currencyPlaceValueStyle == CurrencyValuePlaceStyle.style1) {
        final f = intl.NumberFormat(
            "#${settings.thousandSeparator!}##${settings.thousandSeparator!}###${settings.thousandSeparator!}###${settings.decimalSeparator!}${List.generate(fractionDigits, (index) => "0").join('')}");
        numberUnit = (f.format(amount)).split(".")[0];
        decimalUnit = (f.format(amount)).split(".")[1];
      } else if (currencyPlaceValueStyle == CurrencyValuePlaceStyle.style2) {
        final f = intl.NumberFormat(
            "##${settings.thousandSeparator!}##${settings.thousandSeparator!}##${settings.thousandSeparator!}###${settings.decimalSeparator!}${List.generate(fractionDigits, (index) => "0").join('')}");
        numberUnit = (f.format(amount)).split(".")[0];
        decimalUnit = (f.format(amount)).split(".")[1];
      }
    }

    switch (settings.symbolAlign) {
      case SymbolAlign.left:
        return ['$numberUnit$letter', decimalUnit];
      case SymbolAlign.right:
        return ['$numberUnit$letter', decimalUnit];
      default:
        return ['$numberUnit$letter', decimalUnit];
    }
  }

  /// Parse a formatted string to a number.
  static num parse(String input, CurrencyFormatterSettings settings) {
    String txt = input
        .replaceFirst(settings.thousandSeparator ?? '', '')
        .replaceFirst(settings.decimalSeparator ?? '', '.')
        .replaceFirst(settings.symbol, '')
        .trim();

    String letter = '';
    for (String letter in _letters.values) {
      if (txt.endsWith(letter)) {
        txt = txt.replaceFirst(letter, '');
        letter = letter;
        break;
      }
    }

    return num.parse(txt) *
        _letters.keys.firstWhere((e) => _letters[e] == letter, orElse: () => 1);
  }
}

/// This class contains the formatting settings for a currency.
class CurrencyFormatterSettings {
  /// Symbol of the currency. e.g. '€'
  String symbol;

  /// Whether the symbol is shown before or after the money value, or if it is not shown at all.
  /// e.g. $ 125 ([SymbolAlign.left]) or 125 € ([SymbolAlign.right]).
  SymbolAlign symbolAlign;

  /// Thousand separator. e.g. 1,000,000 (`','`) or 1.000.000 (`'.'`). It can be set to any desired [String].
  /// It defaults to `','` for [SymbolAlign.left] and to `'.'` for [SymbolAlign.right].
  String? thousandSeparator;

  /// Decimal separator. e.g. 9.10 (`'.'`) or 9,10 (`','`). It can be set to any desired [String].
  /// It defaults to `'.'` for [SymbolAlign.left] and to `','` for [SymbolAlign.right].
  String? decimalSeparator;

  CurrencyFormatterSettings(
      {required this.symbol,
      this.symbolAlign = SymbolAlign.left,
      this.thousandSeparator,
      this.decimalSeparator}) {
    thousandSeparator ??= symbolAlign == SymbolAlign.left ? ',' : '.';
    decimalSeparator ??= symbolAlign == SymbolAlign.left ? '.' : ',';
  }

  // Returns the same [CurrencyFormatterSettings] but with some changed parameters.
  CurrencyFormatterSettings copyWith({
    String? symbol,
    SymbolAlign? symbolAlign,
    String? thousandSeparator,
    String? decimalSeparator,
    String? symbolSeparator,
  }) =>
      CurrencyFormatterSettings(
          symbol: symbol ?? this.symbol,
          symbolAlign: symbolAlign ?? this.symbolAlign,
          thousandSeparator: thousandSeparator ?? this.thousandSeparator,
          decimalSeparator: decimalSeparator ?? this.decimalSeparator);
}


lib/currency_decoration.dart
import 'package:flutter/material.dart';

import 'currency_formatter.dart';
import 'currency_symbol.dart';

export 'currency_symbol.dart';

class CurrencyDecoration extends StatelessWidget {
  /// Amount. e.g. 52450.6998
  final double amount;

  /// Symbol of the currency. e.g. '€'
  final String symbol;

  /// If [enforceDecimals] is set to `true`, decimals will be shown even if it is an integer.
  /// e.g. `'$ 5.00'` instead of `'$ 5'`.
  final bool enforceDecimals;

  /// Whether the symbol is shown before or after the money value, or if it is not shown at all.
  /// e.g. $ 125 ([SymbolAlign.left]) or 125 € ([SymbolAlign.right]).
  final SymbolAlign? symbolAlign;

  /// Thousand separator. e.g. 1,000,000 (`','`) or 1.000.000 (`'.'`). It can be set to any desired [String].
  /// It defaults to `','`.
  final String? thousandSeparator;

  /// Decimal separator. e.g. 9.10 (`'.'`) or 9,10 (`','`). It can be set to any desired [String].
  /// It defaults to `'.'`.
  final String? decimalSeparator;

  /// Character(s) between the number and the currency symbol. e.g. $ 9.10 (`' '`) or $9.10 (`''`).
  /// It defaults to a normal space (`' '`).
  final String? symbolSeparator;

  /// If [compact] is `true` the compact form will be used. e.g. `'$ 1.23K'` instead of `'$ 1,230'`.
  final bool? compact;

  /// If [showThousandSeparator] is set to `false`, the thousand separator won't be shown.
  /// e.g. `'$ 1200'`instead of `'$ 1,200'`.
  final bool? showThousandSeparator;

  /// 1.toStringAsFixed(3);  // 1.000
  /// (4321.12345678).toStringAsFixed(3);  // 4321.123
  final int? fractionDigits;

  const CurrencyDecoration(
      {Key? key,
      required this.amount,
      required this.symbol,
      this.symbolAlign = SymbolAlign.left,
      this.thousandSeparator,
      this.decimalSeparator,
      this.symbolSeparator = '',
      this.enforceDecimals = false,
      this.showThousandSeparator = true,
      this.compact,
      this.primaryTextStyle,
      this.currencyValuePlaceStyle = CurrencyValuePlaceStyle.style1,
      this.secondaryTextStyle,
      this.symbolTextStyle,
      this.fractionalStyle,
      this.fractionDigits})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    final CurrencyFormatterSettings currencyFormatterSettings =
        CurrencyFormatterSettings(
            symbol: symbol,
            thousandSeparator: thousandSeparator ?? ',',
            decimalSeparator: decimalSeparator ?? '.');
    final formatted = CurrencyFormatter.format(
        amount, currencyFormatterSettings,
        enforceDecimals: enforceDecimals,
        fractionDigits: fractionDigits ?? 2,
        currencyPlaceValueStyle: currencyValuePlaceStyle,
        showThousandSeparator: showThousandSeparator,
        compact: compact ?? false); // 1.910,93 €

    final DefaultTextStyle defaultTextStyle = DefaultTextStyle.of(context);
    // Primary Text Style
    final TextStyle primeTextStyle = primaryTextStyle == null
        ? defaultTextStyle.style
        : defaultTextStyle.style.merge(primaryTextStyle);
    // Secondary Text Style
    final TextStyle secondTextStyle = primaryTextStyle == null
        ? defaultTextStyle.style
        : primaryTextStyle!.merge(secondaryTextStyle);
    // Symbol Style
    final TextStyle symTextStyle = symbolTextStyle == null
        ? symbolAlign == SymbolAlign.left
            ? primeTextStyle
            : secondTextStyle
        : primeTextStyle.merge(symbolTextStyle);
    return Text.rich(
      key: key,
      TextSpan(text: null, children: <InlineSpan>[
        if (symbolAlign == SymbolAlign.left && symbol.isNotEmpty)
          TextSpan(
            text: "$symbol$symbolSeparator",
            style: symTextStyle,
          ),
        TextSpan(
          text: formatted[0],
          style: primeTextStyle,
        ),
        if (formatted[1].isNotEmpty)
          TextSpan(
            text: "${decimalSeparator ?? '.'}${formatted[1]}",
            style: secondTextStyle,
          ),
        if (symbolAlign == SymbolAlign.right && symbol.isNotEmpty)
          TextSpan(
            text: "$symbolSeparator$symbol",
            style: symTextStyle,
          ),
      ]),
    );
  }

  /// Text style apply on whole units/value (e.g. 52470.596 then text style apply on 52470)
  final TextStyle? primaryTextStyle;

  /// Whole units/value comma style
  final CurrencyValuePlaceStyle? currencyValuePlaceStyle;

  /// Text style apply on fractional part (e.g. 52470.596 then text style apply on 596)
  final TextStyle? secondaryTextStyle;

  /// Text style apply on Currency Symbol
  final TextStyle? symbolTextStyle;

  /// Style on fractional part/value
  final FractionalStyle? fractionalStyle;
}
